crates/sargon/src/signing/collector/signatures_collector_preprocessor.rs

```diff
@@ -0,0 +1,149 @@
+    used_factor_sources: HashSet<FactorSource>,
+) -> IndexSet<FactorSourcesOfKind> {
+    let factors_of_kind: HashMap<FactorSourceKind, IndexSet<FactorSource>> =
+        used_factor_sources
+            .into_iter()
+            .into_grouping_map_by(|x| x.factor_source_kind())
+            .collect::<IndexSet<_>>();
+
+    let mut factors_of_kind = factors_of_kind
+        .into_iter()
+        .map(|(k, v)| (k, v.into_iter().sorted().collect::<IndexSet<_>>()))
+        .collect::<IndexMap<FactorSourceKind, IndexSet<FactorSource>>>();
+
+    factors_of_kind.sort_keys();
+
+    factors_of_kind
+        .into_iter()
+        .map(|(k, v)| {
+            FactorSourcesOfKind::new(k, v)
+                .expect("All factors should be of the same kind, since this is calling iter on a Map, using kind as key. Did you just move around lines of code?")
+        })
+        .collect::<IndexSet<_>>()
+}
+
+impl SignaturesCollectorPreprocessor {
+    pub(super) fn new(transactions: IndexSet<TXToSign>) -> Self {
+        Self { transactions }
+    }
+
+    pub(super) fn preprocess(
+        self,
+        profile_factor_sources: IndexSet<FactorSource>,
+    ) -> (Petitions, IndexSet<FactorSourcesOfKind>) {
+        let transactions = self.transactions;
+        let mut petitions_for_all_transactions =
+            IndexMap::<IntentHash, PetitionForTransaction>::new();
+
+        // We care for only the factor sources which are HD based
+        let mut all_factor_sources_in_profile =
+            HashMap::<FactorSourceIDFromHash, FactorSource>::new();
+        profile_factor_sources.into_iter().for_each(|f| {
+            if let Some(id) = f.factor_source_id().as_hash() {
+                all_factor_sources_in_profile.insert(*id, f);
+            }
+        });
+
+        let mut factor_to_payloads =
+            HashMap::<FactorSourceIDFromHash, IndexSet<IntentHash>>::new();
+
+        let mut used_factor_sources = HashSet::<FactorSource>::new();
+
+        let mut register_factor_in_tx =
+            |id: &FactorSourceIDFromHash, txid: &IntentHash| {
+                if let Some(ref mut txids) = factor_to_payloads.get_mut(id) {
+                    txids.insert(txid.clone());
+                } else {
+                    factor_to_payloads
+                        .insert(*id, IndexSet::just(txid.clone()));
+                }
+
+                assert!(!factor_to_payloads.is_empty());
+
+                let factor_source = all_factor_sources_in_profile
+                    .get(id)
+                    .expect("Should have all factor sources");
+                used_factor_sources.insert(factor_source.clone());
+
+                assert!(!used_factor_sources.is_empty());
+            };
+
+        for transaction in transactions.into_iter() {
+            let mut petitions_for_entities =
+                HashMap::<AddressOfAccountOrPersona, PetitionForEntity>::new();
+
+            for entity in transaction.entities_requiring_auth() {
```

Comment:
better yet, create a constructor on `PetitionForEntity` like this:
```rust
impl PetitionForEntity {
    pub(crate) fn new_with_unsecurified_entity(
        intent_hash: IntentHash, 
        address: AddressOfAccountOrPersona, 
        unsecurified_entity_control: UnsecurifiedEntityControl
    ) -> Self {
        Self::new_unsecurified(intent_hash, address, unsecurified_entity_control.transaction_signing)
    }

      pub(crate) fn new_with_securified_entity(
        intent_hash: IntentHash, 
        address: AddressOfAccountOrPersona, 
        matrix_of_factor_instances: MatrixOfFactorInstances,
        role_selector: RoleSelector
    ) -> Self {
        let role_with_instances = GeneralRoleWithInstances::try_from((matrix_of_factor_instances, role_selector)).unwrap();
        Self::new_with_securified_entity(intent_hash, address, role_with_instances)                       
    }

    pub(crate) fn new_with_entity(intent_hash: IntentHash, entity: AccountOrPersona, if_securified_select_role: RoleSelector) -> Self {
        match entity.entity_security_state() {
            EntitySecurityState::Unsecured { value } => Self::new_with_unsecurified_entity(intent_hash, entity.address(), value)
            EntitySecurityState::Securified { value } => Self::new_with_securified_entity(intent_hash, entity.address(), value.security_structure.matrix_of_factor_instances, if_securified_select_role)
        }
    }
}
```

and then we can update this `for transaction in transactions.into_iter()` into doing this:

```rust
        for transaction in transactions.into_iter() {
            let mut petitions_for_entities =
                HashMap::<AddressOfAccountOrPersona, PetitionForEntity>::new();

            let intent_hash = transaction.intent_hash.clone();
            for entity in transaction.entities_requiring_auth() {
                let petition = PetitionForEntity::new_with_entity(
                    intent_hash.clone(), 
                    entity, 
                    role_selector // <--- SHOULD BE PARAMETER TO THIS FUNCTION `preproccess`
                );
              
       
                petitions.all_factor_instances().iter().for_each(|fi| {
                    register_factor_in_tx(
                        &fi.factor_source_id,
                        intent_hash.clone()
                    );
                });
             
                petitions_for_entities.insert(entity.address(), petition);
            }

            let petition_of_tx = PetitionForTransaction::new(
                transaction.intent_hash.clone(),
                petitions_for_entities,
            );

            petitions_for_all_transactions
                .insert(transaction.intent_hash, petition_of_tx);
        }
```

which is much much much better than the mess I originally wrote. 

I tink think that should work?! Or with minimal changes!